--- a/Makefile
+++ b/Makefile
@@ -127,7 +127,6 @@
 DESTDIR :=
 PREFIX := /usr
 BINDIR := $(PREFIX)/bin
-DOCDIR := $(PREFIX)/share/doc
 MANDIR := $(PREFIX)/share/man
 SYSCONFDIR := /etc
 UNITDIR := /lib/systemd/system
@@ -141,8 +140,6 @@
 	install -D -m 0644 extras/lavinmq.ini $(DESTDIR)$(SYSCONFDIR)/lavinmq/lavinmq.ini
 	install -D -m 0644 extras/lavinmq.service $(DESTDIR)$(UNITDIR)/lavinmq.service
 	install -D -m 0644 extras/lavinmq.sysusers $(DESTDIR)$(SYSUSERSDIR)/lavinmq.conf
-	install -D -m 0644 -t $(DESTDIR)$(DOCDIR)/lavinmq README.md NOTICE
-	install -D -m 0644 CHANGELOG.md $(DESTDIR)$(DOCDIR)/lavinmq/changelog
 	install -d -m 0750 $(DESTDIR)$(SHAREDSTATEDIR)/lavinmq
 
 .PHONY: uninstall
@@ -152,7 +149,6 @@
 	$(RM) $(DESTDIR)$(SYSCONFDIR)/lavinmq/lavinmq.ini
 	$(RM) $(DESTDIR)$(UNITDIR)/lavinmq.service
 	$(RM) $(DESTDIR)$(SYSUSERSDIR)/lavinmq.conf
-	$(RM) -r $(DESTDIR)$(DOCDIR)/lavinmq
 
 .PHONY: rpm
 rpm:
